### PR TITLE
[SYCL-MLIR] Enable C++ 17 support in the SYCL dialect

### PR DESCRIPTION
**Pull request by victor-eds**
Use C++ 17 fold expressions when possible.

Signed-off-by: Victor Perez <victor.perez@codeplay.com>

### CODE REVIEW

**Review by sommerlukas — APPROVED** [review]
LGTM.

**Comment by whitneywhtsang** [comment]
Did you check that it can build cleanly using clang as build compiler too?

File: mlir-sycl/lib/Dialect/IR/SYCLOpsDialect.cpp [hunk 1]
```diff
@@ -87,24 +87,21 @@ bool mlir::sycl::MethodRegistry::registerMethod(mlir::TypeID TypeID,
...
-// If the operation is a SYCL method, register it.
 template <typename T>
-static typename std::enable_if_t<mlir::sycl::isSYCLMethod<T>::value>
-addSYCLMethod(mlir::sycl::MethodRegistry &methods) {
-  const auto TypeID = T::getTypeID();
-  const llvm::StringRef OpName = T::getOperationName();
-  for (llvm::StringRef Name : T::getMethodNames())
-    assert(methods.registerMethod(TypeID, Name, OpName) && "Duplicated method");
+static void addSYCLMethod(mlir::sycl::MethodRegistry &methods) {
+  if constexpr (mlir::sycl::isSYCLMethod<T>::value) {
```
[the remark this answers is not in the document]
Comment (whitneywhtsang) [reply mlir-sycl/lib/Dialect/IR/SYCLOpsDialect.cpp:92]: I prefer early exit, but I don't feel strongly about it. Resolving this conversation.

**Comment by victor-eds** [comment]
> Did you check that it can build cleanly using clang as build compiler too?

I did not. Setting up a clang build. Will report as soon as it finishes.

**Comment by victor-eds** [comment]
> I did not. Setting up a clang build. Will report as soon as it finishes.

Everything's working fine. I'll merge if you are OK with this.

**Review by whitneywhtsang — APPROVED** [review]

**Comment by whitneywhtsang** [comment]
> > I did not. Setting up a clang build. Will report as soon as it finishes.
> 
> Everything's working fine. I'll merge if you are OK with this.

Thanks for checking. Approved.

**Pull request merged**